web/src/message_list.js

```diff
@@ -72,6 +69,10 @@ export class MessageList {
         // the user. Possibly this can be unified in some nice way.
         this.reading_prevented = false;
 
+        // Used to determine whether to preserve the rendered state
+        // when this message list is not the current message list.
+        this.preserve_rendered_state = user_settings.web_home_view === "all_messages" && !this.narrowed;
+
         return this;
     }
 
```

Comment:
So there's one block of logic that I'm not whether it needs attention:

```
    update_muting_and_rerender() {                                                                      
        // For the home message list, we need to re-initialize                                          
        // _all_items for stream muting/topic unmuting from                                             
        // all_messages_data, since otherwise unmuting a previously                                     
        // muted stream won't work.                                                                     
        //                                                                                              
        // "in-home" filter doesn't included muted stream messages, so we                               
        // need to repopulate the message list with all messages to include                             
        // the previous messages in muted streams so that update_items_for_muting works.                
        if (this.data.filter.is_in_home()) {                                                            
            this.data.clear();                                                                          
            this.data.add_messages(all_messages_data.all_messages());                                   
        }                                                                                               
```

I'm not sure that hack still works correctly with the `message_lists.home` logic no longer being deeply tied to `all_messages_data` in the `message_fetch` logic -- it seems like that could very easily result in your not having the currently selected message in view, if that's not present in `all_messages_data`.

It's a messy problem -- we had this performance-favorable model previously that we maintain the full set of messages in `all_messages_data`, and thus always have the ability the rework the data if muting state changes -- but now I think if you scroll up in the view, we'll ask the server for data that we don't end up storing in `all_messages_data`. Hrm.

Reply:
Some version of this problem in theory applies to any search view with `-is:muted` among the filters; but those other views are going to be rare search views where people don't totally expect live-update. (And there's plenty of precedent: The `has:image` search view doesn't live update when someone edits a message to now have an image, and that seems totally OK).

I think there's basically two possible solutions:
- In an unmuting or message-moving event causes some messages to (possibly) now match the combined feed view, and it's cached, we ask the server for full copies of any messages we didn't have copies of locally, a la `maybe_add_narrowed_messages`, as a deferred live-update step.
- Another theory, which I'm not at all sure is a good idea, would be that we arrange that the cacheable `filter.is_in_home` view actually uses the real `all_messages_data as its `.message_list_data`? Since they're both singletons, I think it'd be workable ... but seems messy.

But it would require some preparatory refactoring; this function at least accesses `.items`, not `.all_items` on the `all_messages_data` object:
```
function load_local_messages(msg_data) {                                                                
    // This little helper loads messages into our narrow message                                        
    // data and returns true unless it's visibly empty.  We use this for                                
    // cases when our local cache (all_messages_data) has at least                                      
    // one message the user will expect to see in the new narrow.                                       
                                                                                                        
    const in_msgs = all_messages_data.all_messages();                                                   
    msg_data.add_messages(in_msgs);                                                                     
                                                                                                        
    return !msg_data.visibly_empty();                                                                   
}                                                                                                       
```

Reply:
Yeah, so to be concrete, I think the thing that's worse/regressed is that if you have no unreads and go to the "All messages" view, and just scroll up, you'll do all these fetches from the server that in theory could extend the set of cached messages in `all_messages_data` (as we do with the "load more" button in the Recent Conversations view), but we will no longer do so. 

I think that's probably OK to take no action on for this PR. Opened https://github.com/zulip/zulip/issues/29841 to track this.